src/liboslexec/llvm_util.cpp

```diff
@@ -1751,7 +1753,98 @@ LLVM_Util::InstallLazyFunctionCreator(void* (*P)(const std::string&))
+
+    bool runOnFunction(llvm::Function &F) override {
+        llvm::DominatorTree &dt = getAnalysis<llvm::DominatorTreeWrapperPass>().getDomTree();
+
+        llvm::ValueMap<llvm::BasicBlock*, llvm::DenseSet<int>> bblock_layer_lookup;
+
+        const std::string target_fn = "osl_check_layer_skip_stub";
 
+        // Find all of the stub calls and associate them with their basic block
+        for(auto I = llvm::inst_begin(F), E = llvm::inst_end(F); I != E; ++I) {
+            auto* call_inst = llvm::dyn_cast<llvm::CallInst>(&*I);
+            if (call_inst) {
+                llvm::Function* called = call_inst->getCalledFunction();
+                if (called && called->getName() == target_fn) {
+                    llvm::Value *arg = call_inst->getArgOperand(0);
+                    int layer = llvm::cast<llvm::ConstantInt>(arg)->getSExtValue();
+
+                    llvm::BasicBlock* bblock = I->getParent();
+                    bblock_layer_lookup[bblock].insert(layer);
+                    m_calls_checked++;
+                }
+            }
+        }
+
+        if (bblock_layer_lookup.size() == 0)
+            return false;
+
+        // For each stub call, walk the dominator tree and look for a previous
+        // matching call.
+        std::unordered_set<llvm::CallInst*> delete_queue;
+        for(auto I = llvm::inst_begin(F), E = llvm::inst_end(F); I != E; ++I) {
+            auto* call_inst = llvm::dyn_cast<llvm::CallInst>(&*I);
+            if (call_inst) {
+                llvm::Function* called = call_inst->getCalledFunction();
+                if (called && called->getName() == target_fn) {
+                    llvm::Value *arg = call_inst->getArgOperand(0);
+                    int layer = llvm::cast<llvm::ConstantInt>(arg)->getSExtValue();
+
+                    llvm::BasicBlock* bblock = I->getParent();
+                    auto bbnode = dt.getNode(bblock)->getIDom();
+                    while (bbnode) {
+                        llvm::BasicBlock* candidate_bblock = bbnode->getBlock();
+                        if (bblock_layer_lookup[candidate_bblock].count(layer) > 0) {
+                            delete_queue.insert(call_inst);
+                            break;
+                        }
+                        bbnode = bbnode->getIDom();
+                    }
+                }
+            }
+        }
+
+        int count = delete_queue.size();
+        if (count == 0)
+            return false;
+
+        // Delete all the unnecessary stubs identified above
+
+        // WIP: This appears to be the cause of the performance regression
+        // (Not the dominator tree analysis above)
+        llvm::Value* fake = llvm::ConstantInt::get(F.getContext(), llvm::APInt(32, 1));
+        for (llvm::CallInst* inst : delete_queue) {
+            llvm::BasicBlock::iterator iterator(inst);
+            llvm::ReplaceInstWithValue(inst->getParent()->getInstList(), iterator, fake);
+            m_calls_removed++;
+        }
```

Comment:
This appears to be where the performance hit originates from. It seems like the replacement and subsequent code-folding is pretty simple, so I'm confused why it is taking so much time compared to everything else that happens while codegen'ing a shader (+10-20%).